docs: update template reference variables

4javier opened the pull request:
- Specify `NgForm` gets applied by default on `<form>` elements before the long example using it
- Move the strange (and questioned in a commented line) snippet about undefined ref vars in a standalone paragraph and clarify its meanings (adding the part about directive just because there was something similar already there)
- Extend and modify `*ngFor` example, since in the original that was misleading to think reference variables couldn't be used inside a loop
- Remove two lines stating that with `*ngIf` and `*ngFor` the framework cannot know if a template is ever instantiated (can't see how this relates with the page)
- Add an explanation of assignment of default `$implicit` value to undefined input variables
- Modify template example for template input variable to be a complete ngForOf loop instead of the original poorly intelligible truncated one
- Replace last generic statements about variable namespaces with a more pragmatic and explanatory one concerning the resolution in case of homonymy

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

4javier committed:
docs: update template reference variables

- Specify `NgForm` gets applied by default on `<form>` elements before the long example using it
- Move the strange (and questioned in a commented line) snippet about undefined ref vars in a standalone paragraph and clarify its meanings (adding the part about directive just because there was something similar already there)
- Extend and modify `*ngFor` example, since in the original that was misleading to think reference variable couldn't be used inside a loop
- Remove two lines stating that with `*ngIf` and `*ngFor` the framework cannot know if a template is ever instantiated (can't see how this relate with the page)
- Add an explanation of assignment of default `$implicit` value to undefined input variables
- Modify template example for template input variable to be a complete ngForOf loop instead of the original poor intelligible truncated one
- Replace last generic statements about variable namespaces with a more pragmatic and explanatory one concerning the resolution in case of homonymy

mary-poppins commented:
You can preview a5b456e at https://pr45455-a5b456e.ngbuilds.io/.
You can preview 2959bf9 at https://pr45455-2959bf9.ngbuilds.io/.
You can preview 351047e at https://pr45455-351047e.ngbuilds.io/.

dylhunn commented on aio/content/guide/template-reference-variables.md:
```diff
@@ -46,21 +46,28 @@ In most cases, Angular sets the template variable's value to the element on whic
 
-The `NgForm` directive demonstrates getting a reference to a different value by referencing a directive's `exportAs` name.
-In the following example, the template variable, `itemForm`, appears three times separated by HTML.
+The `NgForm` directive demonstrates getting a reference to a different value by referencing a directive's `exportAs` name.  
```
Replace this paragraph:

The `NgForm` directive is applied by Angular on `<form>` elements. This example demonstrates getting a reference to a different value by referencing a directive's `exportAs` name.

4javier replied:
You're suggesting to remove the following sentence in parenthesis too, right?

dylhunn replied:
Yes, thanks.

4javier replied:
ok

dylhunn commented on aio/content/guide/template-reference-variables.md:
````diff
@@ -127,12 +134,27 @@ Consider the following example that uses `*ngFor`.
+```
 
-With structural directives, such as `*ngFor` or `*ngIf`, there is no way for Angular to know if a template is ever instantiated.
+It's easy to notice how interpolation trying to access property `value` of `ref` occurs outside of referenced element's parent template, making it unreachable.
````
The interpolation trying to access property `ref.value` occurs outside of the referenced element's parent template, making it unreachable.

4javier replied:
Ok.

dylhunn commented on aio/content/guide/template-reference-variables.md:
```diff
@@ -46,21 +46,28 @@ In most cases, Angular sets the template variable's value to the element on whic
+- **Component**: instance of the specific Component class
+- **NgTemplate**: TemplateRef
+
+Keep in mind that just the presence of a directive applied to the element will **not** make the undefined variable default to an instance of that directive, it will still reference the native object.
```
This sentence is unclear, can you rewrite it?

4javier replied:
I added this part just because a similar one was present in the original text. Sounds a bit unnecessary, the only way I could rewrite would be:
_Referencing an element by its directive needs the directive `exportAs` property set as reference value.
In case of unspecified variable value, the reference will return an `HTMLElement`, even if the element has one or more directive applied to itself._

dylhunn replied:
I do think this is better, yes.

dylhunn commented on aio/content/guide/template-reference-variables.md:
```diff
@@ -127,12 +134,27 @@ Consider the following example that uses `*ngFor`.
+It's easy to notice how interpolation trying to access property `value` of `ref` occurs outside of referenced element's parent template, making it unreachable.
 
-As a result, Angular isn't able to access the value and returns an error.
+Moving the interpolation inside the template, will make the variable available referencing the right distinct value for every element the loop will render.
```
Replace:

Moving the interpolation inside the template makes the variable available. Now it references the correct distinct value for each element the loop renders.

4javier replied:
Ok.

dylhunn commented on aio/content/guide/template-reference-variables.md:
````diff
@@ -127,12 +134,27 @@ Consider the following example that uses `*ngFor`.
+</ng-template>
+```
+
+This last snippet will show 2 `<input>` element with their respective value printed just after them
````
Replace:

This snippet shows 2 `<input>` elements, with their respective value printed.

4javier replied:
Ok.

dylhunn commented on aio/content/guide/template-reference-variables.md:
```diff
@@ -127,12 +134,27 @@ Consider the following example that uses `*ngFor`.
+
+This last snippet will show 2 `<input>` element with their respective value printed just after them
 
 ### Accessing a template variable within `<ng-template>`
```
I don't think the discussion of "implicit" makes this section any clearer. Can you revert the changes to this section?

4javier replied:
Where the `$implicit` explanation should belong if not in the paragraph about _template input variables_? 
Is it explained in some other page?
What about the change to the example loop?
And to the namespace stuff?

P.s.: the page title is _Template Reference Variable_, so I think it should be changed or this paragraph moved to a page on its own, but the latter looks undue considering the small length of it.

dylhunn replied:
Hm. I agree that it's just not a great fit for this page. We can leave it though.

dylhunn reviewed (changes requested):
I think the updates to the first section are useful, with some fixes below.

4javier commented:
@dylhunn 
I'm waiting for your opinion about my doubt on those three suggestions before commiting changes to avoid triggering pipeline for many small edits.

dylhunn commented:
@4javier Sorry I didn't see your replies earlier! I responded to your comments.

There's no need to delay pushing -- we can run the CI as many times as we like. Please @ me when you've addressed the remaining comments, thanks! And feel free to bump it if it gets lost in my inbox.

4javier committed:
docs: modified as per dylhunn suggestions

https://github.com/angular/angular/pull/45455#pullrequestreview-925369970

jessicajaniuk commented:
@4javier Please fetch, rebase, and resolve the conflicts. No need to worry about how many times CI runs. Thanks for putting in this effort.

mary-poppins commented:
You can preview eeea0c1 at https://pr45455-eeea0c1.ngbuilds.io/.

4javier committed:
docs: fix to comply with new codeblock guidelines

mary-poppins commented:
You can preview ae63feb at https://pr45455-ae63feb.ngbuilds.io/.

4javier commented:
@dylhunn @jessicajaniuk 
I rebased and fixed conflicts.
Last commit is to update my added codeblocks to use `<code-example>` elements as per https://github.com/angular/angular/commit/42289f25c69ec7a36d2e3ef424d32f77bc7cd19a
IMHO: 
- having to escape special characters is really inconvenient
- GitHub preview ignores `<codeblock>` elements, making the tool useless to quickly check if they've been used correctly

dylhunn commented on aio/content/guide/template-reference-variables.md:
```diff
@@ -47,21 +47,29 @@ In most cases, Angular sets the template variable's value to the element on whic
+- **NgTemplate**: TemplateRef
+
+Referencing an element by its directive needs the directive `exportAs` property set as reference value.
+In case of unspecified variable value, the reference will return an `HTMLElement`, even if the element has one or more directive applied to itself.
```
grammar: "In case of an unspecified"

4javier replied:
Done.

dylhunn reviewed (approved):
Looks good, just one little fix.

4javier committed:
docs: fix grammar

mary-poppins commented:
You can preview 8712f08 at https://pr45455-8712f08.ngbuilds.io/.

jessicajaniuk commented:
This PR was merged into the repository by commit a6ab3f7beb2c16caff1433ba224f6e96f873b6ec.
